zmq_addon.hpp

```diff
@@ -408,190 +408,115 @@ class multipart_t
-        multipart_t head("One", 3);
-        head.addstr("Two");
-        assert(head.size() == 2);
-
-        multipart_t tail("One-hundred");
-        tail.pushstr("Ninety-nine");
-        assert(tail.size() == 2);
-
-        multipart_t tmp(message_t("Fifty", 5));
-        assert(tmp.size() == 1);
-
-        multipart_t mid = multipart_t::create(49.0f);
-        mid.append(std::move(tmp));
-        assert(mid.size() == 2);
-        assert(tmp.empty());
-
-        multipart_t merged(std::move(mid));
-        merged.prepend(std::move(head));
-        merged.append(std::move(tail));
-        assert(merged.size() == 6);
-        assert(head.empty());
-        assert(tail.empty());
-
-        ok = merged.send(output);
-        assert(merged.empty());
-        assert(ok);
-
-        multipart_t received(input);
-        assert(received.size() == 6);
-
-        str = received.popstr();
-        assert(received.size() == 5);
-        assert(str == "One");
-
-        str = received.popstr();
-        assert(received.size() == 4);
-        assert(str == "Two");
-
-        num = received.poptyp<float>();
-        assert(received.size() == 3);
-        assert(num == 49.0f);
-
-        str = received.popstr();
-        assert(received.size() == 2);
-        assert(str == "Fifty");
-
-        str = received.popstr();
-        assert(received.size() == 1);
-        assert(str == "Ninety-nine");
-
-        str = received.popstr();
-        assert(received.empty());
-        assert(str == "One-hundred");
-
-        return 0;
-    }
-
 private:
     // Disable implicit copying (moving is more efficient)
     multipart_t(const multipart_t& other) ZMQ_DELETED_FUNCTION;
     void operator=(const multipart_t& other) ZMQ_DELETED_FUNCTION;
 };  // class multipart_t
 
-#endif  // ZMQ_HAS_RVALUE_REFS
-
 inline std::ostream& operator<<(std::ostream& os, const multipart_t& msg)
 {
     return os << msg.str();
 }
 
+#endif  // ZMQ_HAS_RVALUE_REFS
+
+#if defined(ZMQ_BUILD_DRAFT_API) && defined(ZMQ_CPP11) && defined(ZMQ_HAVE_POLLER)
+    class active_poller_t
+    {
+    public:
+        active_poller_t () = default;
+        ~active_poller_t () = default;
+
+        active_poller_t(const active_poller_t&) = delete;
+        active_poller_t &operator=(const active_poller_t&) = delete;
+
+        active_poller_t(active_poller_t&& src) = default;
+        active_poller_t &operator=(active_poller_t&& src) = default;
+
+        using handler_t = std::function<void(short)>;
+
+        void add (zmq::socket_t &socket, short events, handler_t handler)
+        {
+            auto it = decltype (handlers)::iterator {};
+            auto inserted = bool {};
+            std::tie(it, inserted) = handlers.emplace (static_cast<void*>(socket), std::make_shared<handler_t> (std::move (handler)));
+            try 
+            {
+                base_poller.add (socket, events, inserted && *(it->second) ? it->second.get() : nullptr);
+                need_rebuild |= inserted;
```

Comment:
depending on default initialization here makes it only harder to read.

Reply:
In order to avoid an unnecessary rebuild?
IIRC it was like this before, just something like
```
if (inserted)
{
  need_rebuild = true;
  ...
}
```

Reply:
You not avoiding here anything. If you are after `base_poller.add` you have to rebuild regardless of `inserted`.

Reply:
Hm, I think add will throw necessarily if inserted is false, so the need_rebuild line will only be reached if inserted is true anyway. So... it could be changed...